pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+		}
+
+		/// Allows root to update the buyout limit.
+		///
+		/// Parameters
+		///
+		/// - `origin`: Origin must be root.
+		/// - `limit`: New buyout limit. If None, then buyouts are not limited.
+		///
+		/// Emits `BuyoutLimitUpdated` event when successful.
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::update_buyout_limit())]
+		pub fn update_buyout_limit(
+			origin: OriginFor<T>,
+			limit: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			match limit {
+				Some(limit) => BuyoutLimit::<T>::put(limit),
+				None => BuyoutLimit::<T>::kill(),
+			}
+			Self::deposit_event(Event::<T>::BuyoutLimitUpdated { limit });
+			Ok(().into())
+		}
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Attempt to exchange native token to native token
+		WrongAssetToBuyout,
+		/// Buyout limit exceeded for the current period
+		BuyoutLimitExceeded,
+		/// One of transacted currencies is missing price information
+		NoPrice,
+		/// The treasury balance is too low for an operation
+		InsufficientTreasuryBalance,
+		/// The account balance is too low for an operation
+		InsufficientAccountBalance,
+		/// Exchange failed
+		ExchangeFailure,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Buyout event
+		Buyout {
+			who: AccountIdOf<T>,
+			buyout_amount: BalanceOf<T>,
+			asset: CurrencyIdOf<T>,
+			exchange_amount: BalanceOf<T>,
+		},
+		/// Buyout limit updated event
+		BuyoutLimitUpdated { limit: Option<BalanceOf<T>> },
+	}
+
+	/// Stores limit amount user could by for a period.
+	/// When `None` - buyouts are not limited
+	#[pallet::storage]
+	pub type BuyoutLimit<T: Config> = StorageValue<_, BalanceOf<T>, OptionQuery>;
+
+	/// Stores amount of buyouts (amount, block number of last buyout)
+	#[pallet::storage]
+	pub type Buyouts<T: Config> =
+		StorageMap<_, Blake2_128Concat, AccountIdOf<T>, (BalanceOf<T>, u32), ValueQuery>;
+}
+
+impl<T: Config> Pallet<T> {
+	/// Ensures that buyout limit is not exceeded for the current buyout period
+	fn ensure_buyout_limit_not_exceeded(
+		account_id: &AccountIdOf<T>,
+		buyout_amount: BalanceOf<T>,
+	) -> DispatchResult {
+		if let Some(buyout_limit) = BuyoutLimit::<T>::get() {
+			let buyout_period = T::BuyoutPeriod::get();
+			// Get current block number
+			let now = <frame_system::Pallet<T>>::block_number().saturated_into::<u32>();
+			let current_period = now
+				.checked_div(buyout_period)
+				.and_then(|n| Some(n.saturating_mul(buyout_period)))
+				.unwrap_or_default();
+			let (mut buyouts, last_buyout) = Buyouts::<T>::get(account_id);
+
+			if !buyouts.is_zero() && last_buyout < current_period {
+				buyouts = Default::default();
+				Buyouts::<T>::insert(account_id, (buyouts, now));
+			};
+
+			ensure!(
+				buyouts
+					.saturated_into::<u128>()
+					.saturating_add(buyout_amount.saturated_into::<u128>()) <=
+					buyout_limit.saturated_into::<u128>(),
```

Comment:
Do we have to do the `saturated_into()` calls? Is there a way to safely add `Balance` types without casting here?

Reply:
You're right, we don't really need the casting. Updated the code 👍🏼